pulp_ansible/app/tasks/collections.py

```diff
@@ -674,21 +674,46 @@ async def _add_namespace(self, name):
             pass
         else:
             namespace = parse_metadata(result)
-            namespace.pop("pulp_href", None)
-            url = namespace.pop("avatar_url", None)
-            da = (
-                [
-                    DeclarativeArtifact(
-                        Artifact(sha256=namespace["avatar_sha256"]),
-                        url=url,
-                        remote=self.remote,
-                        relative_path=f"{name}-avatar",
-                        deferred_download=False,
-                    )
-                ]
-                if url
-                else None
-            )
+            links = namespace.get("links", None)
+
+            # clean up the galaxy API for pulp
+            if links:
+                namespace["links"] = {x["name"]: x["url"] for x in links}
+
+            for key in ("pulp_href", "groups", "id", "related_fields"):
+                namespace.pop(key, None)
+
+            avatar_sha256 = namespace.get("avatar_sha256", None)
+            url = namespace.get("avatar_url", None)
+
+            if url:
+                if avatar_sha256:
+                    a = Artifact(sha256=avatar_sha256)
+                else:
+                    a = Artifact()
+                logo = DeclarativeArtifact(
+                    a,
+                    url=url,
+                    remote=self.remote,
+                    relative_path=f"{name}-avatar",
+                    deferred_download=False,
+                )
+
+                await logo.download()
+
+                # If the remote doesn't support avatar_sha256, save the avatar_url to the
+                # database. This will cause the namespace metadata has to be calculated using
+                # the url instead of the sha, which will allow it to be computed correctly for
+                # syncs against galaxy_ng
+                if avatar_sha256:
+                    namespace.pop("avatar_url", None)
+                else:
+                    namespace["avatar_sha256"] = logo.artifact.sha256
```

Comment:
Do we care if the upstream `avatar_url` changes the avatar it is pointing to? I'm pretty sure in that scenario we wouldn't properly update the namespace. The sync might even break.

Reply:
@gerrod3 do you mean the image the avatar_url points to changes?

Reply:
If that's what you mean, it won't download the new image if something else hasn't been changed on the namespace. I don't think it will break, because when the avatar_url is set on the ns model, that gets used for the metadata sha

Reply:
Yes, in this case NamespaceMetadata would break on a resync because the metadata_sha256 wouldn't change (the url didn't change), but the avatar sha256 did. The NamespaceMetadata wouldn't be updated since the primary key didn't change, but since we downloaded the new avatar and put it into a declaritive artifact, the sync pipeline will update the artifact for the NamespaceMetadata. Then there will be a mismatch for the avatarsha256 and the actual avatar being stored/served. 

Two options (that I see): 

1. Don't download the logo when `avatar_url` is used. Have Pulp always redirect to original link. Pros - clients that fetch the avatar will always get the most up to date one. Cons - no longer storing avatar locally.
2. Include both `avatar_url` and `avatar_sha256` in `metadata_sha256` calculation. This would desync the metadatasha256 from the remote and Pulp, but Pulp would now properly create new AnsibleMetadata when the avatar image changes.

Reply:
looks like you're right